bofire/utils/torch_tools.py

```diff
@@ -188,24 +188,57 @@ def constrained_objective2botorch(
         objective (BotorchConstrainedObjective): The objective that should be transformed.
 
     Returns:
-        Tuple[List[Callable[[Tensor], Tensor]], List[float]]: List of callables that can be used by botorch for setting up the constrained objective, and
-            list of the corresponding botorch eta values.
+        Tuple[List[Callable[[Tensor], Tensor]], List[float], int]: List of callables that can be used by botorch for setting up the constrained objective,
+            list of the corresponding botorch eta values, final index used by the method (to track for categorical variables)
     """
     assert isinstance(
         objective, ConstrainedObjective
     ), "Objective is not a `ConstrainedObjective`."
     if isinstance(objective, MaximizeSigmoidObjective):
-        return [lambda Z: (Z[..., idx] - objective.tp) * -1.0], [
-            1.0 / objective.steepness
-        ]
+        return (
+            [lambda Z: (Z[..., idx] - objective.tp) * -1.0],
+            [1.0 / objective.steepness],
+            idx + 1,
+        )
     elif isinstance(objective, MinimizeSigmoidObjective):
-        return [lambda Z: (Z[..., idx] - objective.tp)], [1.0 / objective.steepness]
+        return (
+            [lambda Z: (Z[..., idx] - objective.tp)],
+            [1.0 / objective.steepness],
+            idx + 1,
+        )
     elif isinstance(objective, TargetObjective):
-        return [
-            lambda Z: (Z[..., idx] - (objective.target_value - objective.tolerance))
-            * -1.0,
-            lambda Z: (Z[..., idx] - (objective.target_value + objective.tolerance)),
-        ], [1.0 / objective.steepness, 1.0 / objective.steepness]
+        return (
+            [
+                lambda Z: (Z[..., idx] - (objective.target_value - objective.tolerance))
+                * -1.0,
+                lambda Z: (
+                    Z[..., idx] - (objective.target_value + objective.tolerance)
+                ),
+            ],
+            [1.0 / objective.steepness, 1.0 / objective.steepness],
+            idx + 1,
+        )
+    elif isinstance(objective, ConstrainedCategoricalObjective):
+        # The output of a categorical objective has final dim `c` where `c` is number of classes
+        # Pass in the expected acceptance probability and perform an inverse sigmoid to atain the original probabilities
+        return (
+            [
+                lambda Z: torch.log(
+                    torch.clamp(
+                        1
+                        / (
+                            Z[..., idx : idx + len(objective.desirability)]
+                            * torch.tensor(objective.desirability).to(**tkwargs)
+                        ).sum(-1)
+                        - 1,
+                        min=eps,
+                        max=1 - eps,
+                    )
+                )
+            ],
+            [objective.eta],
```

Comment:
I would opt for removing the `objective.eta` and make sure that the eta which is given here is always 1 for identity, or?

Reply:
I can do this, but will it break functionality elsewhere? Is eta always required for type Objective?

Reply:
Should not be a problem.